dot/node.go

```
@@ -41,9 +47,86 @@ type Node struct {
+		net *network.Service) error
+	createBlockVerifier(st *state.Service) (*babe.VerificationManager, error)
+	createDigestHandler(lvl log.Level, st *state.Service) (*digest.Handler, error)
+	createCoreService(cfg *Config, ks *keystore.GlobalKeystore, st *state.Service, net *network.Service,
+		dh *digest.Handler) (*core.Service, error)
+	createGRANDPAService(cfg *Config, st *state.Service, dh *digest.Handler, ks keystore.Keystore,
+		net *network.Service, telemetryMailer telemetry.Client) (*grandpa.Service, error)
+	newSyncService(cfg *Config, st *state.Service, fg gsync.FinalityGadget, verifier *babe.VerificationManager,
+		cs *core.Service, net *network.Service, telemetryMailer telemetry.Client) (*gsync.Service, error)
+	createBABEService(cfg *Config, st *state.Service, ks keystore.Keystore, cs *core.Service,
+		telemetryMailer telemetry.Client) (*babe.Service, error)
+	createSystemService(cfg *types.SystemInfo, stateSrvc *state.Service) (*system.Service, error)
+	createRPCService(params rpcServiceSettings) (*rpc.HTTPServer, error)
+}
+
+type nodeBuilder struct{}
+
+// NodeInitialized returns true if, within the configured data directory for the
+// node, the state database has been created and the genesis data has been loaded
+func NodeInitialized(basepath string) bool {
+	nodeInstance := nodeBuilder{}
+	return nodeInstance.nodeInitialised(basepath)
+}
+
+func (nodeBuilder) nodeInitialised(basepath string) bool {
+	// check if key registry exists
+	registry := path.Join(basepath, utils.DefaultDatabaseDir, "KEYREGISTRY")
+
+	_, err := os.Stat(registry)
+	if os.IsNotExist(err) {
+		logger.Debug("node has not been initialised from base path " + basepath +
+			": failed to locate KEYREGISTRY file in data directory")
+
+		return false
+	}
+
+	// initialise database using data directory
+	db, err := utils.SetupDatabase(basepath, false)
+	if err != nil {
+		logger.Debugf("failed to create database from base path %s: %s", basepath, err)
+		return false
```

All these logs in this function should return an error to `NodeInitialized`, which can itself wrap the error with i.e. `fmt.Errorf("cannot initialise node from base path %s: %w", basepath, err)` and then `NodeInitialized` can return the error or just log it out.

The context of the base path and of the `failed to initialise node` should belong to the caller of `nodeInitialised`

---

Good idea, I've updated this.